reconcile/utils/oc_map.py

```diff
@@ -0,0 +1,317 @@
+
+    @property
+    def cluster_admin_automation_token(self) -> Optional[HasSecret]:
+        ...
+
+    @property
+    def disable(self) -> Optional[HasDisable]:
+        ...
+
+
+class HasNamespace(Protocol):
+    cluster: HasCluster
+    cluster_admin: Optional[bool]
+
+
+class OCMap:
+    """OCMap gets a GraphQL query results list as input
+    and initiates a dictionary of OC clients per cluster.
+
+    The input must contain either 'clusters' or 'namespaces', but not both.
+
+    In case a cluster does not have an automation token
+    the OC client will be initiated with a OCLogMessage.
+    """
+
+    def __init__(
+        self,
+        clusters: Optional[Sequence[HasCluster]] = None,
+        clusters_untyped: Optional[Sequence[MutableMapping[Any, Any]]] = None,
+        namespaces: Optional[Sequence[HasNamespace]] = None,
+        namespaces_untyped: Optional[Sequence[MutableMapping[Any, Any]]] = None,
+        integration: str = "",
+        e2e_test: str = "",
+        settings_untyped: Optional[Mapping[Any, Any]] = None,
+        secret_reader: Optional[SecretReaderBase] = None,
+        internal: Optional[bool] = None,
+        use_jump_host: bool = True,
+        thread_pool_size: int = 1,
+        init_projects: bool = False,
+        init_api_resources: bool = False,
+        cluster_admin: bool = False,
+    ):
+        self._oc_map: dict[str, Union[OCDeprecated, OCLogMsg]] = {}
+        self._privileged_oc_map: dict[str, Union[OCDeprecated, OCLogMsg]] = {}
+        self._calling_integration = integration
+        self._calling_e2e_test = e2e_test
+        self._secret_reader = secret_reader
+        self._internal = internal
+        self._use_jump_host = use_jump_host
+        self._thread_pool_size = thread_pool_size
+        self._init_projects = init_projects
+        self._init_api_resources = init_api_resources
+        self._lock = Lock()
+        self._jh_ports: dict[str, int] = {}
+        self._settings_dict = settings_untyped
+        self._jumphosts_dict: dict[Any, Any] = {}
+
+        if clusters and namespaces:
+            raise KeyError("expected only one of clusters or namespaces.")
+        elif clusters and clusters_untyped:
+            for cluster_dict in clusters_untyped:
+                self._jumphosts_dict[cluster_dict.get("name")] = cluster_dict.get(
+                    "jumpHost"
+                )
+            threaded.run(
+                self.init_oc_client,
+                clusters,
+                self._thread_pool_size,
+                privileged=cluster_admin,
+            )
+        elif namespaces and namespaces_untyped:
+            for ns_dict in namespaces_untyped:
+                cluster_d = ns_dict.get("cluster")
+                if cluster_d:
+                    self._jumphosts_dict[cluster_d.get("name")] = cluster_d.get(
+                        "jumpHost"
+                    )
+            clusters_dict = {}
+            privileged_clusters: dict[str, HasCluster] = {}
```

Comment:
Do we need to use dict here? I can only find usage of `.values()`. If not then can be simplified with list like
```python
unprivileged_clusters = [ns_info.cluster for ns_info in namespaces]
privileged_clusters = unprivileged_clusters if cluster_admin else [ns_info.cluster for ns_info in namespaces if ns_info.cluster_admin]

```

Reply:
@fishi0x01 i can relate a lot to your `class ClusterParameters` proposal. having minimal input is nice in many ways, e.g. lots of integration don't need to have access to unnecessary schemas (`allowedSchemas`) hence run less frequent during PR checks. also, it would be a lot easier to write tests.

Reply:
@hemslo good point. The idea was to copy/paste as much as possible from the existing logic and make it type-compatible with as less changes as possible. I.e., I didn't do any optimization on the logic itself.

Reply:
If possible, I'd like to address these optimizations in a final polishing PR, once it was confirmed that the types are working w/o breaking anything.

Reply:
Due to introduction of `OCConnectionParameters` class I had to change the `OCMap` `__init__` function and applied your above mentioned suggestion.